p2panda-rs/src/schema/schema.rs

```diff
@@ -0,0 +1,426 @@
+use std::collections::BTreeMap;
+use std::fmt;
+
+use crate::hash::Hash;
+use crate::message::{Message, MessageFields, MessageValue};
+use crate::schema::SchemaError;
+
+use cddl::lexer::Lexer;
+use cddl::parser::Parser;
+#[cfg(not(target_arch = "wasm32"))]
+use cddl::validate_cbor_from_slice;
+#[cfg(not(target_arch = "wasm32"))]
+use cddl::validator::cbor;
+
+/// CDDL types
+#[derive(Clone, Debug, Copy)]
+#[allow(missing_docs)]
+pub enum Type {
+    Bool,
+    Int,
+    Float,
+    Tstr,
+    Relation,
+}
+
+/// CDDL schema type string formats
+impl fmt::Display for Type {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let cddl_type = match self {
+            Type::Bool => "bool",
+            Type::Int => "int",
+            Type::Float => "float",
+            Type::Tstr => "tstr",
+            Type::Relation => "tstr .regexp \"[0-9a-fa-f]{132}\"",
+        };
+        write!(f, "{}", cddl_type)
+    }
+}
+
+/// CDDL field types
+#[derive(Clone, Debug)]
+pub enum Field {
+    String(String),
+    Type(Type),
+    Group(Group),
+    Vector(Group),
+    Struct(Group),
+    Table(Group),
+    TableType(Type),
+    Choice(Group),
+}
+
+/// Format each different data type into a schema string.
+impl fmt::Display for Field {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Field::String(s) => write!(f, "\"{}\"", s),
+            Field::Type(cddl_type) => write!(f, "{}", cddl_type),
+            Field::Group(group) => write!(f, "{}", group),
+            Field::Vector(group) => write!(f, "[* {} ]", format!("{}", group)),
+            Field::Struct(group) => write!(f, "{{ {} }}", format!("{}", group)),
+            Field::Table(group) => write!(f, "{{ + tstr => {{ {} }} }}", group),
+            Field::TableType(value_type) => write!(f, "{{ + tstr => {} }}", value_type),
+            Field::Choice(group) => write!(f, "&{}", group),
+        }
+    }
+}
+
+/// Struct for building and representing CDDL groups.
+/// CDDL uses groups to define reuseable data structures
+/// they can be merged into schema or used in Vectors, Tables and Structs
+#[derive(Clone, Debug)]
+pub struct Group {
+    name: String,
+    fields: BTreeMap<String, Field>,
+}
+
+impl Group {
+    /// Create a new CDDL group
+    pub fn new(name: String) -> Self {
+        Self {
+            name,
+            fields: BTreeMap::new(),
+        }
+    }
+
+    /// Add an Field to the group.
+    pub fn add_field(&mut self, key: &str, field_type: Field) {
+        self.fields.insert(key.to_owned(), field_type);
+    }
+}
+
+impl fmt::Display for Group {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let map = &self.fields;
+        write!(f, "( ")?;
+        for (count, value) in map.iter().enumerate() {
+            // For every element except the first, add a comma.
+            if count != 0 {
+                write!(f, ", ")?;
+            }
```

Comment:
Does this help here? https://doc.rust-lang.org/std/primitive.slice.html#method.join

Reply:
I'm not sure, isn't this following a particular pattern for implementing the `fmt::Display` trait?

Reply:
It's not particular only for formatting but one could write it like this maybe?

`map.map(|value| format!("{}: {}", value.0, value.1)).join(", ")` (this is pseudo Rust code)